### PR TITLE
Dynamic transformer

### PR DESCRIPTION
**Pull request by jlamypoirier**
# ✨ Description

Adjust the rotary embeddings, peft and normalization layers to use the new dynamic classes. Do some cleanup and refactoring for rotary embeddings. Add an option to disable normalization layers because why not.

## 🔍 Type of change

Select all that apply:

- [ ] 🐛 **Bug fix** (non-breaking change that addresses a specific issue)
- [x] 🚀 **New feature** (non-breaking change that adds functionality)
- [ ] ⚠️ **Breaking change** (a change that could affect existing functionality)
- [ ] 📈 **Performance improvement/optimization** (improves speed, memory usage, or efficiency)
- [x] 🛠️ **Code refactor** (non-functional changes that improve code readability, structure, etc.)
- [ ] 📦 **Dependency bump** (updates dependencies, including Dockerfile or package changes)
- [ ] 📝 **Documentation change** (updates documentation, including new content or typo fixes)
- [ ] 🔧 **Infrastructure/Build change** (affects build process, CI/CD, or dependencies)

### CODE REVIEW

File: fast_llm/layers/transformer/rotary/rotary.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+    def _get_frequencies(self, sequence_length: int, kv_channels: int, device="cuda") -> torch.Tensor:
+        return super()._get_frequencies(sequence_length, kv_channels, device) * self._config.attention_factor
+
+    def _get_angle_scales(self, kv_channels: int, device="cuda") -> torch.Tensor:
+        scales = super()._get_angle_scales(kv_channels, device)
+        # TODO: max_position_embeddings or original_context_length?
+        # see https://huggingface.co/deepseek-ai/DeepSeek-V3/blob/main/modeling_deepseek.py#L304
+        low = max(self._get_correction(self._config.beta_slow, kv_channels), 0)
+        high = min(self._get_correction(self._config.beta_fast, kv_channels), kv_channels - 1)
```
[the remark this answers is not in the document]
Comment (RaymondLi0) [reply fast_llm/layers/transformer/rotary/rotary.py:191]: Hmm, well the original implementation uses 
- `beta_fast` to compute `low` 
- `beta_slow` to compute `high`

File: fast_llm/layers/transformer/rotary/rotary.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    def _linear_ramp_factor(self, min, max, dim):
+        if min == max:
+            max += 0.001  # Prevent singularity
+
+        linear_func = (torch.arange(dim, dtype=torch.float32) - min) / (max - min)
+        ramp_func = torch.clamp(linear_func, 0, 1)
+        return ramp_func
+
+    def _get_correction(self, beta: float, dim: int) -> float:
+        return math.floor(
```
Comment (RaymondLi0) [inline fast_llm/layers/transformer/rotary/rotary.py:210]: original implementation uses `floor` for `low`, but `ceil` for `high`.

File: fast_llm/layers/transformer/rotary/rotary.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jlamypoirier) [reply fast_llm/layers/transformer/rotary/rotary.py:191]: You're right, sorry for the confusion

File: fast_llm/layers/transformer/config.py [hunk 3]
```diff
@@ -194,66 +160,50 @@ class TransformerPeftConfig(PeftConfig):
     )
```
Comment (oleksost) [inline fast_llm/layers/transformer/config.py:160]: Not sure this is the right place to raise this concern, but I think its better to not have layer freezing as part of PEFT, and keep only a single layer freezing mechanism (e.g. currently using lr sclaing for different components and layers).  

Hence, if user wants to use PEFT (e.g. LoRA) in conjunction with layer freezing, they might do it by explicitly freezing the layers using `lr_scale` parameter. 

This way we can keep the LoRA/PEFT logic simple and disentangle parameter freezing from PEFT.

File: fast_llm/layers/common/config.py [hunk 4]
```diff
@@ -120,27 +136,47 @@ def _from_dict(
...
+
+    @property
+    def module_class(self):
+        from fast_llm.layers.common.normalization import RMSNorm
+
+        return RMSNorm
 
 
 @config_class()
 class PeftConfig(BaseModelConfig):
```
Comment (oleksost) [inline fast_llm/layers/common/config.py:162]: no default handling (from_dict)?

Comment (jlamypoirier) [reply fast_llm/layers/common/config.py:162]: We don't really use that one and it doesn't have a registry, these are in `TransformerPeftConfig` instead.

File: fast_llm/layers/transformer/config.py (hunk 3 above)
Comment (jlamypoirier) [reply fast_llm/layers/transformer/config.py:160]: That's a fair point, but LoRA is almost always used together with freezing, so it's really more convenient to do it together. Also I don't think the arbitrary lr scaling parameters is a good long term solution, and anyway there is still only one freezing mechanism in the background, so it's not too bad.
